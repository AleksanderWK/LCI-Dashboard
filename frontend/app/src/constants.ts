// Frequency in Hz (cycles per second)
export const FREQUENCY = 2;

// Maximum time interval to show on live charts in seconds
export const LIVE_CHART_RANGE = 30;

export enum Variable {
    CognitiveLoad = "cl",
    PerceivedDifficulty = "pd",
    Familiarity = "fam",
    InformationProcessingIndex = "ipi",
    PhysiologicalArousal = "pa",
    Engagement = "eng",
    PhysiologicalStress = "ps",
    EmotionalRegulation = "er",
    MotionStability = "ms",
    EnergySpentFatigue = "esf",
    EducationalSpecificEmotions = "ese"
}

export enum Device {
    EyeTracker = "Eye Tracker",
    Wristband = "Wristband",
    VideoBody = "Video of Body",
    VideoFace = "Video of Face"
}

export enum EyeTrackingDevice {
    Mobile = "Mobile",
    Stationary = "Stationary"
}

export interface MMDVariableInfo {
    name: string;
    device: Device;
    description: string;
    maxValue?: number;
    calculationTime?: number;
    enabled: boolean;
}

export const MMDVariables: {[key in Variable]: MMDVariableInfo} = {
    [Variable.CognitiveLoad]: {
        name: "Cognitive Load",
        device: Device.EyeTracker,
        description:
            "Cognitive load refers to the amount of information that the working memory of a person can hold at once. In an educational setting, it is the effort needed for a learner to learn new information.",
        calculationTime: 30,
        enabled: true
    },
    [Variable.PerceivedDifficulty]: {
        name: "Perceived Difficulty",
        device: Device.EyeTracker,
        description:
            "Perceived difficulty is the perception of the ease or difficulty of performing in an educational context.",
        maxValue: 100,
        enabled: true
    },
    [Variable.Familiarity]: {
        name: "Familiarity",
        device: Device.EyeTracker,
        description: "",
        enabled: false
    },
    [Variable.InformationProcessingIndex]: {
        name: "Information Processing Index",
        device: Device.EyeTracker,
        description:
            "The ratio of local (100) and global (0) information processing. Staring at one point is local processing, while rapidly scanning a larger area is global processing.",
        maxValue: 100,
        enabled: true
    },
    [Variable.PhysiologicalArousal]: {
        name: "Physiological Arousal",
        device: Device.Wristband,
        description:
            "Physiological Arousal is how strong the current emotions are. It is calculated by the average EDA in a time interval of 8 seconds.",
        calculationTime: 8,
        enabled: true
    },
    [Variable.Engagement]: {
        name: "Engagement",
        device: Device.Wristband,
        description:
            "Engagement is a measure of involvement and enthusiasm for the activity. It is calculated by the area under the graph of tonic EDA, in a time interval of 4 seconds.",
        calculationTime: 4,
        enabled: true
    },
    [Variable.PhysiologicalStress]: {
        name: "Physiological Stress",
        device: Device.Wristband,
        description:
            "Physioloigcal Stress is indicated by LF/HF, the relative strengths of the low frequency and high frequency waves of the heart pulse signal.",
        calculationTime: 300,
        enabled: true
    },
    [Variable.EmotionalRegulation]: {
        name: "Emotional Regulation",
        device: Device.Wristband,
        description: "",
        enabled: false
    },
    [Variable.MotionStability]: {
        name: "Motion Stability",
        device: Device.VideoBody,
        description: "",
        enabled: false
    },
    [Variable.EnergySpentFatigue]: {
        name: "Energy Spent/Fatigue",
        device: Device.VideoBody,
        description:
            "The amount of energy spent by the student, also an indicator of fatigue. This is determined by calculating the jerk, the third derivative of the position skeletal body.",
        calculationTime: 3,
        enabled: true
    },
    [Variable.EducationalSpecificEmotions]: {
        name: "Educational-specific Emotions",
        device: Device.VideoFace,
        description: "Emotions that the student is experiencing in an educational setting.",
        enabled: true
    }
};

export const emotionsColorMapper: {[key: string]: string} = {
    boredom: "#F15025",
    frustration: "#191919",
    confusion: "#FFE45E",
    delight: "#6782B7"
};

export const emotionsIndexMapper: {[key: string]: number} = {
    boredom: 0,
    frustration: 1,
    confusion: 2,
    delight: 3
};

export const months = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December"
];
