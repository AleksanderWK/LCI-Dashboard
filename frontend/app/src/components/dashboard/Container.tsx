import React, {useRef, useState} from "react";
import {
    Card,
    CardContent,
    createStyles,
    IconButton,
    makeStyles,
    SvgIcon,
    SvgIconProps,
    Theme,
    Typography
} from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import {Variable} from "../../constants";
import Tooltip from "./Tooltip";
import Menu from "./Menu";
import LineChart from "./LineChart";

function ResizeIcon(props: SvgIconProps): JSX.Element {
    return (
        <SvgIcon {...props}>
            <path d="M24 24H19.2V19.2H24V24ZM24 14.4H19.2V9.6H24V14.4ZM14.4 24H9.6V19.2H14.4V24ZM14.4 14.4H9.6V9.6H14.4V14.4ZM4.8 24H0V19.2H4.8V24ZM24 4.8H19.2V0H24V4.8Z" />
        </SvgIcon>
    );
}

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        card: {
            maxWidth: 500,
            position: "relative"
        },
        header: {
            display: "grid",
            gridTemplateColumns: "1fr 56px",
            gap: theme.spacing(2),
            marginBottom: theme.spacing(2)
        },
        menu: {
            display: "grid",
            gridTemplateColumns: "24px 24px",
            gap: theme.spacing(1)
        },
        iconButton: {
            padding: 0
        },
        infoIcon: {
            cursor: "default"
        },
        resizeIcon: {
            position: "absolute",
            right: "4px",
            bottom: "4px",
            width: "10px",
            height: "10px",
            "&:hover": {
                cursor: "nw-resize"
            }
        }
    })
);

interface Props {
    variable: Variable;
}

export default function Container(props: Props): JSX.Element {
    const classes = useStyles();

    const [isDetailedView, setIsDetailedView] = useState<boolean>(true);

    const [menuOpen, setMenuOpen] = useState<boolean>(false);

    const menuAnchorElement = useRef<HTMLDivElement | null>(null);

    return (
        <Card variant="outlined" className={classes.card}>
            <CardContent>
                <div className={classes.header}>
                    <Typography variant="h2" noWrap={true}>
                        {props.variable}
                    </Typography>

                    <div className={classes.menu} ref={menuAnchorElement}>
                        <Tooltip variable={props.variable}>
                            <IconButton
                                aria-label="info"
                                disableFocusRipple={true}
                                disableRipple={true}
                                disableTouchRipple={true}
                                className={`${classes.iconButton} ${classes.infoIcon}`}
                            >
                                <InfoOutlinedIcon color="action" />
                            </IconButton>
                        </Tooltip>
                        <IconButton
                            aria-label="settings"
                            className={classes.iconButton}
                            onClick={() => setMenuOpen(true)}
                        >
                            <MoreVertIcon color="action" />
                        </IconButton>
                    </div>

                    <Menu
                        open={menuOpen}
                        anchorEl={menuAnchorElement.current}
                        isDetailedView={isDetailedView}
                        onShowMore={() => setIsDetailedView(true)}
                        onShowLess={() => setIsDetailedView(false)}
                        onRemoveView={() => null}
                        onMenuClose={() => setMenuOpen(false)}
                    />
                </div>

                {isDetailedView ? <LineChart /> : <div>Insert value component here</div>}
            </CardContent>
            <ResizeIcon className={classes.resizeIcon} color="action" />
        </Card>
    );
}
