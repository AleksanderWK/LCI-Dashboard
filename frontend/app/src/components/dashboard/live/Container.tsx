import React, {useRef, useState} from "react";
import {MMDVariables, Variable} from "../../../constants";
import Menu from "./Menu";
import LineChart from "./LineChart";
import Numeric from "./Numeric";
import ContainerCard from "../ContainerCard";
import Tooltip from "../Tooltip";
import {makeStyles, Theme, createStyles, IconButton, Typography} from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import CalculatingIndicator from "./CalculatingIndicator";
import {useRecoilState, useRecoilValue} from "recoil";
import {
    selectedAllSessionVariableState,
    selectedSessionActiveContainersState,
    selectedSessionDataLengthVariableState,
    selectedSessionIdState,
    sessionVariableDataState
} from "../../../state/session";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        header: {
            display: "grid",
            gridTemplateColumns: "1fr 56px",
            gap: theme.spacing(2)
        },
        menu: {
            display: "grid",
            gridTemplateColumns: "24px 24px",
            gap: theme.spacing(1)
        },
        iconButton: {
            padding: 0
        },
        infoIcon: {
            cursor: "default"
        }
    })
);

interface Props {
    variable: Variable;
    display: "line" | "numeric";
    studentName?: string;
    id?: number;
}

export default function Container(props: Props): JSX.Element {
    const classes = useStyles();

    const [containerOptions, setContainerOptions] = useRecoilState(selectedSessionActiveContainersState);

    const [menuOpen, setMenuOpen] = useState<boolean>(false);

    const menuAnchorElement = useRef<HTMLDivElement | null>(null);

    const selectedSessionId = useRecoilValue(selectedSessionIdState);
    const selectedAllSessionsVariable = useRecoilValue(selectedAllSessionVariableState);
    const dataLength = props.id
        ? useRecoilValue(sessionVariableDataState([props.variable, props.id])).length
        : useRecoilValue(selectedSessionDataLengthVariableState(props.variable));

    function setChartType(chartType: "line" | "numeric"): void {
        const newContainerOptions = {...containerOptions};
        newContainerOptions[props.variable] = {...containerOptions[props.variable], display: chartType};
        setContainerOptions(newContainerOptions);
    }

    function removeView(): void {
        const newContainerOptions = {...containerOptions};
        newContainerOptions[props.variable] = {...containerOptions[props.variable], active: false};
        setContainerOptions(newContainerOptions);
    }

    return (
        <ContainerCard>
            <>
                <div className={classes.header}>
                    <Typography variant="h2" noWrap={true}>
                        {selectedSessionId ? MMDVariables[props.variable].name : props.studentName}
                    </Typography>

                    <div className={classes.menu} ref={menuAnchorElement}>
                        <Tooltip variable={props.variable}>
                            <IconButton
                                aria-label="info"
                                disableFocusRipple={true}
                                disableRipple={true}
                                disableTouchRipple={true}
                                className={`${classes.iconButton} ${classes.infoIcon}`}
                            >
                                <InfoOutlinedIcon color="action" />
                            </IconButton>
                        </Tooltip>
                        <IconButton
                            aria-label="settings"
                            className={classes.iconButton}
                            onClick={() => setMenuOpen(true)}
                        >
                            <MoreVertIcon color="action" />
                        </IconButton>
                    </div>

                    <Menu
                        open={menuOpen}
                        anchorEl={menuAnchorElement.current}
                        isDetailedView={containerOptions[props.variable].display === "line"}
                        onShowMore={() => setChartType("line")}
                        onShowLess={() => setChartType("numeric")}
                        onRemoveView={() => removeView()}
                        onMenuClose={() => setMenuOpen(false)}
                    />
                </div>
                {selectedSessionId != null ? (
                    <>
                        {MMDVariables[props.variable].calculationTime && dataLength === 0 ? (
                            <CalculatingIndicator variable={props.variable} />
                        ) : containerOptions[props.variable].display === "line" ? (
                            <LineChart variable={props.variable} />
                        ) : (
                            <Numeric variable={props.variable} />
                        )}
                    </>
                ) : (
                    <>
                        {selectedAllSessionsVariable &&
                        MMDVariables[selectedAllSessionsVariable as Variable].calculationTime &&
                        dataLength === 0 ? (
                            <CalculatingIndicator variable={props.variable} />
                        ) : props.display === "line" ? (
                            <LineChart variable={props.variable} />
                        ) : (
                            <Numeric variable={props.variable} />
                        )}
                    </>
                )}
            </>
        </ContainerCard>
    );
}
