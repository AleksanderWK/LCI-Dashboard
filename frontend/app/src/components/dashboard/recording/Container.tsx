import {useRef, useState} from "react";
import {createStyles, IconButton, makeStyles, Theme, Typography} from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import {MMDVariables, Variable} from "../../../constants";
import Tooltip from "../Tooltip";
import Menu from "./Menu";
import LineChart from "./LineChart";
import ContainerCard from "../ContainerCard";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        header: {
            display: "grid",
            gridTemplateColumns: "1fr 56px",
            gap: theme.spacing(2)
        },
        menu: {
            display: "grid",
            gridTemplateColumns: "24px 24px",
            gap: theme.spacing(1)
        },
        iconButton: {
            padding: 0
        },
        infoIcon: {
            cursor: "default"
        }
    })
);

interface Props {
    variable: Variable;
}

export default function Container(props: Props): JSX.Element {
    const classes = useStyles();

    const [menuOpen, setMenuOpen] = useState<boolean>(false);

    const menuAnchorElement = useRef<HTMLDivElement | null>(null);

    return (
        <ContainerCard>
            <>
                <div className={classes.header}>
                    <Typography variant="h2" noWrap={true}>
                        {MMDVariables[props.variable].name}
                    </Typography>

                    <div className={classes.menu} ref={menuAnchorElement}>
                        <Tooltip variable={props.variable}>
                            <IconButton
                                aria-label="info"
                                disableFocusRipple={true}
                                disableRipple={true}
                                disableTouchRipple={true}
                                className={`${classes.iconButton} ${classes.infoIcon}`}
                            >
                                <InfoOutlinedIcon color="action" />
                            </IconButton>
                        </Tooltip>
                        <IconButton
                            aria-label="settings"
                            className={classes.iconButton}
                            onClick={() => setMenuOpen(true)}
                        >
                            <MoreVertIcon color="action" />
                        </IconButton>
                    </div>

                    <Menu
                        open={menuOpen}
                        anchorEl={menuAnchorElement.current}
                        onRemoveView={() => null}
                        onMenuClose={() => setMenuOpen(false)}
                    />
                </div>

                <LineChart variable={props.variable} />
            </>
        </ContainerCard>
    );
}
