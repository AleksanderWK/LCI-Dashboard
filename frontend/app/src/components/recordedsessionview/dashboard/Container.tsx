import {useRef, useState} from "react";
import {createStyles, IconButton, makeStyles, Theme, Typography} from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import {MMDVariables, Variable} from "../../../constants";
import Tooltip from "../../common/Tooltip";
import Menu from "./Menu";
import LineChart from "./LineChart";
import ContainerCard from "../../common/ContainerCard";
import XRangeChart from "./XRangeChart";
import {
    selectedRecordedSessionLayoutState,
    selectedRecordingActiveContainersState
} from "../../../state/recordedSession";
import {useSetRecoilState} from "recoil";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        header: {
            display: "grid",
            gridTemplateColumns: "1fr 56px",
            gap: theme.spacing(2)
        },
        menu: {
            display: "grid",
            gridTemplateColumns: "24px 24px",
            gap: theme.spacing(1)
        },
        iconButton: {
            padding: 0
        },
        infoIcon: {
            cursor: "default"
        }
    })
);

interface Props {
    variable: Variable;
}

export default function Container(props: Props): JSX.Element {
    const classes = useStyles();

    const [menuOpen, setMenuOpen] = useState<boolean>(false);

    const menuAnchorElement = useRef<HTMLDivElement | null>(null);

    const setSelectedRecordingActiveContainers = useSetRecoilState(selectedRecordingActiveContainersState);
    const setSelectedRecordingLayout = useSetRecoilState(selectedRecordedSessionLayoutState);

    function removeContainer(): void {
        setSelectedRecordingActiveContainers((prevState) => {
            const newState = [...prevState];
            return newState.filter((item) => item != props.variable);
        });

        // Remove from layout
        setSelectedRecordingLayout((prevValue) => {
            const layout = [...prevValue];
            return layout.filter((item) => item.i != props.variable);
        });
    }

    return (
        <ContainerCard>
            <>
                <div className={classes.header}>
                    <Typography variant="h2" noWrap={true}>
                        {MMDVariables[props.variable].name}
                    </Typography>

                    <div className={`${classes.menu} noDrag`} ref={menuAnchorElement}>
                        <Tooltip variable={props.variable}>
                            <IconButton
                                aria-label="info"
                                disableFocusRipple={true}
                                disableRipple={true}
                                disableTouchRipple={true}
                                className={`${classes.iconButton} ${classes.infoIcon}`}
                            >
                                <InfoOutlinedIcon color="action" />
                            </IconButton>
                        </Tooltip>
                        <IconButton
                            aria-label="settings"
                            className={classes.iconButton}
                            onClick={() => setMenuOpen(true)}
                        >
                            <MoreVertIcon color="action" />
                        </IconButton>
                    </div>

                    <Menu
                        open={menuOpen}
                        anchorEl={menuAnchorElement.current}
                        onRemoveView={() => removeContainer()}
                        onMenuClose={() => setMenuOpen(false)}
                    />
                </div>
                {props.variable == Variable.EducationalSpecificEmotions ? (
                    <XRangeChart variable={props.variable} />
                ) : (
                    <LineChart variable={props.variable} />
                )}
            </>
        </ContainerCard>
    );
}
