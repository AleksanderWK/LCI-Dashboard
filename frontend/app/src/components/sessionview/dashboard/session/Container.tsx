import React, {useRef, useState} from "react";
import {MMDVariables, Variable} from "../../../../constants";
import Menu from "../Menu";
import LineChart from "../LineChart";
import Numeric from "../Numeric";
import ContainerCard from "../../../common/ContainerCard";
import Tooltip from "../../../common/Tooltip";
import {makeStyles, Theme, createStyles, IconButton, Typography} from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import CalculatingIndicator from "../CalculatingIndicator";
import {useRecoilState, useRecoilValue, useSetRecoilState} from "recoil";
import {selectedSessionDataLengthVariableState, selectedSessionIdState} from "../../../../state/session";
import XRangeChart from "../XRangeChart";
import {viewState, selectedSessionActiveContainersState, selectedSessionLayoutState} from "../../../../state/dashboard";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        header: {
            display: "grid",
            gridTemplateColumns: "1fr 56px",
            gap: theme.spacing(2)
        },
        menu: {
            display: "grid",
            gridTemplateColumns: "24px 24px",
            gap: theme.spacing(1)
        },
        iconButton: {
            padding: 0
        },
        infoIcon: {
            cursor: "default"
        }
    })
);

interface Props {
    variable: Variable;
}

export default function Container(props: Props): JSX.Element {
    const classes = useStyles();

    const dataLength = useRecoilValue(selectedSessionDataLengthVariableState(props.variable));

    const selectedSessionId = useRecoilValue(selectedSessionIdState);
    const [view, setView] = useRecoilState(viewState([selectedSessionId, props.variable]));
    const setActiveContainers = useSetRecoilState(selectedSessionActiveContainersState);
    const setSelectedSessionLayout = useSetRecoilState(selectedSessionLayoutState);

    const [menuOpen, setMenuOpen] = useState<boolean>(false);

    const menuAnchorElement = useRef<HTMLDivElement | null>(null);

    function removeContainer(): void {
        setActiveContainers((prevState) => {
            const newState = [...prevState];
            return newState.filter((item) => item != props.variable);
        });

        // Remove from layout
        setSelectedSessionLayout((prevValue) => {
            const layout = [...prevValue];
            return layout.filter((item) => item.i != props.variable);
        });
    }

    return (
        <ContainerCard>
            <>
                <div className={classes.header}>
                    <Typography variant="h2" noWrap={true}>
                        {MMDVariables[props.variable].name}
                    </Typography>

                    <div className={`${classes.menu} noDrag`} ref={menuAnchorElement}>
                        <Tooltip variable={props.variable}>
                            <IconButton
                                aria-label="info"
                                disableFocusRipple={true}
                                disableRipple={true}
                                disableTouchRipple={true}
                                className={`${classes.iconButton} ${classes.infoIcon}`}
                            >
                                <InfoOutlinedIcon color="action" />
                            </IconButton>
                        </Tooltip>
                        <IconButton
                            aria-label="settings"
                            className={classes.iconButton}
                            onClick={() => setMenuOpen(true)}
                        >
                            <MoreVertIcon color="action" />
                        </IconButton>
                    </div>

                    <Menu
                        open={menuOpen}
                        anchorEl={menuAnchorElement.current}
                        isDetailedView={view === "chart"}
                        onShowMore={() => setView("chart")}
                        onShowLess={() => setView("numeric")}
                        onRemoveView={() => removeContainer()}
                        onMenuClose={() => setMenuOpen(false)}
                    />
                </div>
                {MMDVariables[props.variable].calculationTime && dataLength === 0 ? (
                    <CalculatingIndicator variable={props.variable} />
                ) : view === "chart" && props.variable != Variable.EducationalSpecificEmotions ? (
                    <LineChart variable={props.variable} />
                ) : view === "chart" ? (
                    <XRangeChart variable={props.variable} />
                ) : (
                    <Numeric variable={props.variable} />
                )}
            </>
        </ContainerCard>
    );
}
