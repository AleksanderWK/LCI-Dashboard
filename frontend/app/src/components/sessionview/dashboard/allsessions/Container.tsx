import {useRef, useState} from "react";
import {MMDVariables, Variable} from "../../../../constants";
import Menu from "../Menu";
import ContainerCard from "../../../common/ContainerCard";
import Tooltip from "../../../common/Tooltip";
import {makeStyles, Theme, createStyles, IconButton, Typography} from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import {useRecoilState, useRecoilValue, useSetRecoilState} from "recoil";
import {sessionIdsState} from "../../../../state/session";
import {selectedSessionActiveContainersState, selectedSessionLayoutState, viewState} from "../../../../state/dashboard";
import ContainerItem from "./ContainerItem";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        header: {
            display: "grid",
            gridTemplateColumns: "1fr 56px",
            gap: theme.spacing(2)
        },
        menu: {
            display: "grid",
            gridTemplateColumns: "24px 24px",
            gap: theme.spacing(1)
        },
        iconButton: {
            padding: 0
        },
        infoIcon: {
            cursor: "default"
        },
        items: {
            display: "grid",
            gridTemplateColumns: "1fr",
            gap: theme.spacing(3)
        }
    })
);

interface Props {
    variable: Variable;
}

export default function Container(props: Props): JSX.Element {
    const classes = useStyles();

    const sessionIds = useRecoilValue(sessionIdsState);

    const [view, setView] = useRecoilState(viewState([null, props.variable]));

    const setActiveContainers = useSetRecoilState(selectedSessionActiveContainersState);
    const setSelectedSessionLayout = useSetRecoilState(selectedSessionLayoutState);

    const [menuOpen, setMenuOpen] = useState<boolean>(false);

    const menuAnchorElement = useRef<HTMLDivElement | null>(null);

    function removeContainer(): void {
        setActiveContainers((prevState) => {
            const newState = [...prevState];
            return newState.filter((item) => item != props.variable);
        });

        // Remove from layout
        setSelectedSessionLayout((prevValue) => {
            const layout = [...prevValue];
            return layout.filter((item) => item.i != props.variable);
        });
    }

    return (
        <ContainerCard>
            <>
                <div className={classes.header}>
                    <Typography variant="h2" noWrap={true}>
                        {MMDVariables[props.variable].name}
                    </Typography>

                    <div className={`${classes.menu} noDrag`} ref={menuAnchorElement}>
                        <Tooltip variable={props.variable}>
                            <IconButton
                                aria-label="info"
                                disableFocusRipple={true}
                                disableRipple={true}
                                disableTouchRipple={true}
                                className={`${classes.iconButton} ${classes.infoIcon}`}
                            >
                                <InfoOutlinedIcon color="action" />
                            </IconButton>
                        </Tooltip>
                        <IconButton
                            aria-label="settings"
                            className={classes.iconButton}
                            onClick={() => setMenuOpen(true)}
                        >
                            <MoreVertIcon color="action" />
                        </IconButton>
                    </div>

                    <Menu
                        open={menuOpen}
                        anchorEl={menuAnchorElement.current}
                        isDetailedView={view === "chart"}
                        onShowMore={() => setView("chart")}
                        onShowLess={() => setView("numeric")}
                        onRemoveView={() => removeContainer()}
                        onMenuClose={() => setMenuOpen(false)}
                    />
                </div>

                <div className={classes.items}>
                    {sessionIds.map((id) => (
                        <ContainerItem key={id} id={id} variable={props.variable} view={view} />
                    ))}
                </div>
            </>
        </ContainerCard>
    );
}
