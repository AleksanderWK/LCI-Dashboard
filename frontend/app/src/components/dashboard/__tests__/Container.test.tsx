import React from "react";
import ReactDOM from "react-dom";
import {render, screen, waitFor} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {MMDVariables, Variable} from "../../../constants";
import Container from "../Container";

it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Container variable={Variable.CognitiveLoad} />, div);
});

it("renders title", () => {
    render(<Container variable={Variable.CognitiveLoad} />);
    expect(screen.getByText(Variable.CognitiveLoad)).toBeInTheDocument();
});

it("tooltip displays variable info", async () => {
    render(<Container variable={Variable.CognitiveLoad} />);
    userEvent.hover(screen.getByLabelText("info"));
    await waitFor(() => {
        expect(screen.getByText(MMDVariables[Variable.CognitiveLoad].description)).toBeInTheDocument();
        expect(screen.getByText(MMDVariables[Variable.CognitiveLoad].device, {exact: false})).toBeInTheDocument();
    });
});

// Update this when state management is in place
/*
it("menu toggle between show more/less", async () => {
    render(<Container variable={Variable.CognitiveLoad} />);

    userEvent.click(screen.getByLabelText("settings"));
    await screen.findAllByTestId("dropdown-menu").then(async (elements) => {
        expect(elements[elements.length - 1]).toBeVisible();
    });

    userEvent.click(screen.getAllByLabelText("show less")[screen.getAllByLabelText("show less").length - 1]);
    await waitFor(() => {
        expect(screen.getAllByLabelText("show more").pop()).toBeInTheDocument();
    });

    userEvent.click(screen.getAllByLabelText("show more")[screen.getAllByLabelText("show more").length - 1]);
    await waitFor(() => {
        expect(screen.getAllByLabelText("show less").pop()).toBeInTheDocument();
    });
});
*/
