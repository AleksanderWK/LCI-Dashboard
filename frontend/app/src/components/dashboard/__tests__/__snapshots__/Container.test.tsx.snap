// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container matches snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="makeStyles-header-2"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 MuiTypography-noWrap"
          >
            Cognitive Load
          </h2>
          <div
            class="makeStyles-menu-3"
          >
            <button
              aria-label="info"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-4 makeStyles-infoIcon-5"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorAction"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="settings"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-4"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorAction"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div>
          Insert chart component here
        </div>
      </div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root makeStyles-resizeIcon-6 MuiSvgIcon-colorAction"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M24 24H19.2V19.2H24V24ZM24 14.4H19.2V9.6H24V14.4ZM14.4 24H9.6V19.2H14.4V24ZM14.4 14.4H9.6V9.6H14.4V14.4ZM4.8 24H0V19.2H4.8V24ZM24 4.8H19.2V0H24V4.8Z"
        />
      </svg>
    </div>
  </div>
  <div
    aria-hidden="true"
    class="MuiPopover-root"
    data-testid="dropdown-menu"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px; visibility: hidden;"
  >
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation3 MuiPaper-rounded"
      style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiMenu-list MuiList-padding"
        role="menu"
        tabindex="0"
      >
        <ul
          class="MuiList-root makeStyles-menuList-11 MuiList-dense MuiList-padding"
          role="menu"
          tabindex="0"
        >
          <li
            aria-disabled="false"
            aria-label="show less"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-dense MuiMenuItem-dense MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14zM7 9h5v1H7z"
                />
              </svg>
            </div>
            <span
              class="MuiTypography-root"
            >
              Show less
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            aria-label="remove view"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-dense MuiMenuItem-dense MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16 9v10H8V9h8m-1.5-6h-5l-1 1H5v2h14V4h-3.5l-1-1zM18 7H6v12c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7z"
                />
              </svg>
            </div>
            <span
              class="MuiTypography-root"
            >
              Remove view
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </li>
        </ul>
      </ul>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
